Render html before the image

epinter opened the pull request:
With this change, the graph url is generated and returned to the browser as a
base64 string inside the imgsrc. Now the user doesn't have to wait the image
from grafana to view information about service/host, the image is sent
in a separated http connection.

The image is downloaded and sent to browser when init function is called (before
any header or html is sent by Icinga framework).

Mikesch-mp commented:
Do i understand it right that the image url will be visible (encoded or not)?
The proxy mode right now was for companies request that users, other then the admins, can never see the real url to grafana, so they can not watch things they are not allowed to see.

epinter commented:
The image url will be: /monitoring/service/show?host=host&service=service&grafanaimage=<a_long_base64_string>. This base64 string has the encoded url the module generates, don't have username or password.

When browser tries to get the image, the module is accessed again passing the image url in grafanaimage parameter. This parameter is read inside the init function of the module, when the curl occurs, image is sent to browser then module exits.

Mikesch-mp commented:
Hmmm thats enough for most that the proxy mode is not usable anymore because a use could decode the url of the server and try to do something with it (dont ask me why, but some big companies dont want to show it to users because of security ....).

epinter commented:
May this could be another mode ? I can strip the hostname too ...
I only implemented this because Icinga takes a lot of time (1s) to load a service page when the graph is being rendered.

epinter commented:
Carsten,
I removed the grafanahost from url sent to browser, and implemented encryption using the php-openssl extension. The extension is detected and used automatically (every distribution has openssl as default, as far as I can tell). The encryption key is generated using information from config (including password).

epinter commented:
Updated to show error messages as png image, in case of a curl failure.

friesoft commented:
Any news on this one? Icingaweb2 feels pretty slow in proxy mode which is the only one we can use atm..

Mikesch-mp commented:
To make it clear. I will never commit this pr, because it makes things complicated and not supportable anymore. If you read the manual and listen to the professionals of InfluxDB (and or Graphite) you would know you server needs a lot of RAM. My InfluxDB Server has 4 cores and 64GB ram, the configuration is tuned to this environment and it takes  less then a second to load a servicepage inside Icingaweb2.
And for the "bild" readers, if you need a porsche and but bought only a Smart, dont blame Smart beccause its to slow.

Regards,
Carsten

Mikesch-mp commented:
Hello @epinter ,

i finaly found a solution and a safe way to obey all user restriction. So images get loaded by reference, this make pageload very fast. will be released with 1.2.1, but soon to test in master. See #126
